werkzeug/test.py

```diff
@@ -339,6 +345,19 @@ def __init__(self, path='/', base_url=None, query_string=None,
                     else:
                         self.form.setlistdefault(key).append(value)
 
+        if json:
+            if not isinstance(json, dict):
```

Comment:
When I originally started this feature I was trying to emulate the `requests` library request argument that took a dictionary.

Did you have more types in mind? I can only think of strings, which if you've already encoded your payload, just send it as `data` like you have been.